### PR TITLE
bpo-43080: pprint for dataclass instances

### PR DESCRIPTION
**Pull request by LewisGaul**
https://bugs.python.org/issue43080

See also: https://github.com/python/cpython/pull/14318

### CODE REVIEW

**Comment by LewisGaul** [comment]
I've realised this doesn't respect user-defined `__repr__()` methods. I wasn't able to use the existing mechanism of comparing the IDs of `__repr__` in the `_dispatch` dict as dynamically created `__repr__`s all have different IDs. The best solution to this I can think of is to check `object.__class__.__repr__.__module__` - this will be `'dataclasses'` if defined by the dataclass machinery. I'm not sure how watertight this solution is though?

File: Lib/pprint.py
Comment (ericvsmith) [inline Lib/pprint.py]: I think you should ignore fields where `f.repr` is false.

File: Lib/pprint.py
Comment (ericvsmith) [inline Lib/pprint.py]: I don't think this test is needed either.

File: Lib/pprint.py
Comment (ericvsmith) [inline Lib/pprint.py]: This test isn't needed, is it?

**Review by ericvsmith — COMMENTED** [review]
I'm still not convinced this is an awesome idea, but see my review comments anyway.

**Comment by LewisGaul** [comment]
> I'm still not convinced this is an awesome idea, but see my review comments anyway.

I'm happy to address the comments, but if we're not expecting it to be merged shall we just close the PR? I would like to see this feature make it into core, but I won't be offended if you're against it overall, and better not to waste time fixing stuff up if it's not going in anyway! :)

File: Lib/pprint.py
Comment (LewisGaul) [reply Lib/pprint.py]: Agreed. Similarly, I've opted not to do special formatting for dataclasses created with `repr=False` (otherwise we'd need to move the check for whether the object is a dataclass above the check for whether the default repr is short enough, which feels like the wrong change to be making).

**Comment by LewisGaul** [comment]
> I've realised this doesn't respect user-defined `__repr__()` methods. I wasn't able to use the existing mechanism of comparing the IDs of `__repr__` in the `_dispatch` dict as dynamically created `__repr__`s all have different IDs.

I've added a failing testcase to track this issue.

@ericvsmith Do you have any thoughts on how best to tackle this? You suggested the following on the issue, is this approach worth pursuing here?

> Adding an attribute on the __repr__ (and other methods?) signifying that they were generated would let us distinguish them.
>
> Say, checking getattr(__repr__, '__is_generated__', False), or similar.

**Comment by ericvsmith** [comment]
For the `__repr__` check, how about examining `__repr__.__qualname__`  and checking for `__create_fn__.<locals>.__repr__`? I think that's a sufficiently unlikely string that it could be used. Although it's an implementation detail and probably a fragile test, I'd be okay with using it inside the stdlib. Or we could change the name of `__create_fn__` to `__dataclasses_create_fn__` to reinforce the fact that this is a dataclasses internal.

**Comment by LewisGaul** [comment]
> For the `__repr__` check, how about examining `__repr__.__qualname__` and checking for `__create_fn__.<locals>.__repr__`? I think that's a sufficiently unlikely string that it could be used. Although it's an implementation detail and probably a fragile test, I'd be okay with using it inside the stdlib. Or we could change the name of `__create_fn__` to `__dataclasses_create_fn__` to reinforce the fact that this is a dataclasses internal.

I see what you're suggesting when trying it out on 3.8, but it seems the qualname was 'fixed' in https://github.com/python/cpython/pull/22155, so this approach won't work on master. Any other ideas/preferences? :)

**Comment by ericvsmith** [comment]
> I see what you're suggesting when trying it out on 3.8, but it seems the qualname was 'fixed' in #22155, so this approach won't work on master. Any other ideas/preferences? :)

You could do the same with `__repr__.__wrapped__`:

```>>> C.__repr__.__wrapped__
<function __create_fn__.<locals>.__repr__ at 0x0000022FF4824670>
```

Check that `__wrapped__` exists and contains `"__create_fn__"`.

Again, it's super fragile, but since both modules are in the stdlib and will be tested to work together, I'm not so concerned.

**Comment by ericvsmith** [comment]
> Again, it's super fragile, but since both modules are in the stdlib and will be tested to work together, I'm not so concerned.

And you can blame me in a comment if you want!

**Comment by LewisGaul** [comment]
@ericvsmith this should be good for second review/merge now :)

**Comment by ericvsmith** [comment]
@LewisGaul : I'm planning on fixing a bug (which I can't find right now) dealing with repr of recursive dataclasses. I'm going to hold off on this until that one is merged, at which point we can do the same fix here. But if you haven't heard from me in a week or so, please ping me again.

**Comment by ericvsmith** [comment]
@LewisGaul : Can you add support and tests for recursive dataclasses? See https://github.com/python/cpython/blob/3f3d82b84823eb28abeedf317bbe107bbe7f6492/Lib/test/test_dataclasses.py#L3296

**Comment by LewisGaul** [comment]
@ericvsmith I've added handling for recursive dataclasses. Note that this required some special-casing (to avoid backwards incompatibility of formatting other recursive objects), where the existing handling in the pprint module would have given the following:
```
dataclass4(a=<Recursion on dataclass4 with id=2119600742448>)
```

**Review by ericvsmith — CHANGES_REQUESTED** [review]
One last request for a cyclical test.

**Comment by LewisGaul** [comment]
I have made the requested changes; please review again.

I don't think the CI failures are related (`test_ssl` failure on macOS, docs build failure).

**Review by ericvsmith — APPROVED** [review]

**Comment by LewisGaul** [comment]
Thanks for taking the time to review @ericvsmith :) Is there anything else I need to do before this can be merged?

**Comment by ericvsmith** [comment]
@LewisGaul :  No, nothing else. I'll try and finish up my final review and push of this tonight. Thanks for your patience.

**Pull request merged**
